src/csharp/Grpc.Core/GrpcEnvironment.cs

```diff
@@ -360,12 +361,25 @@ internal static string GetCoreVersionString()
 
         internal static void GrpcNativeInit()
         {
+            if (!IsNativeShutdownAllowed && nativeInitCounter.Count > 0)
+            {
+                // Normally grpc_init and grpc_shutdown calls should come in pairs (C core does reference counting),
```

Comment:
LGTM. 

Just another idea though: if going with this lock-free approach, I wonder if some kind of a "lock free once" algorithm would be better here. Also, the mutex vs. lock free counter thing seems like an interesting comparison to run/validate in the C# microbenchmarks.

Reply:
yeah I can't really point to one. 

Actually, can we not do something like:

```
{
   if (nativeInitCounter.Increment() == 0) {
     ... call grpc_init()
   }
   return;
}
```

?

Reply:
I also thought of that but I think it's incorrect:
1. thread 1  calls  Increment, gets 0, then yields 
2. thread 2 calls Increment, gets 1 and thus skips grpc_init
3. thread 2 does something assuming grpc_init has already been invoked, but that' not the case
4. thread 1 finally invokes grpc_init, but it's too late

In another words, we must not increment before grpc_init has returned.